internal/telemetry/collector/segmentio.go

```diff
@@ -0,0 +1,491 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package collector holds different implementation of telemetry data collectors.
+package collector
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	libsveltosv1beta1 "github.com/projectsveltos/libsveltos/api/v1beta1"
+	"github.com/segmentio/analytics-go/v3"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	clusterapiv1 "sigs.k8s.io/cluster-api/api/v1beta1" // TODO: switch to v1beta2 everywhere
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kcmv1 "github.com/K0rdent/kcm/api/v1beta1"
+	"github.com/K0rdent/kcm/internal/utils/pointer"
+)
+
+type SegmentIO struct {
+	segmentCl    analytics.Client
+	crCl         client.Client
+	childScheme  *runtime.Scheme
+	expBackoff   backoff.BackOff
+	childFactory func([]byte, *runtime.Scheme) (client.Client, error) // for test mocks
+	concurrency  int
+}
+
+type (
+	nodeDataAccumulator struct {
+		nodeInfos     []map[string]string
+		totalCPU      uint64
+		totalMemory   uint64
+		totalGPUNodes uint
+		count         uint
+	}
+
+	podDataAccumulator struct {
+		nodeInfos       *[]map[string]string
+		nodeName2Idx    map[string]int
+		podsWithGPUReqs uint
+	}
+)
+
+func NewSegmentIO(segmentClient analytics.Client, crClient client.Client, concurrency int) (*SegmentIO, error) {
+	s := runtime.NewScheme()
+
+	for _, f := range []func(*runtime.Scheme) error{
+		corev1.AddToScheme,
+		metav1.AddMetaToScheme,
+		kubevirtv1.AddToScheme,
+		appsv1.AddToScheme,
+	} {
+		if err := f(s); err != nil {
+			return nil, fmt.Errorf("failed to add to scheme: %w", err)
+		}
+	}
+
+	return &SegmentIO{
+		expBackoff:   backoff.NewExponentialBackOff(backoff.WithInitialInterval(500*time.Millisecond), backoff.WithMaxElapsedTime(10*time.Second)),
+		segmentCl:    segmentClient,
+		crCl:         crClient,
+		childScheme:  s,
+		concurrency:  concurrency,
+		childFactory: defaultClientFactory,
+	}, nil
+}
+
+func (t *SegmentIO) Collect(ctx context.Context) error {
+	l := ctrl.LoggerFrom(ctx)
+
+	mgmt := new(kcmv1.Management)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.Get(ctx, client.ObjectKey{Name: kcmv1.ManagementName}, mgmt)
+	}); err != nil {
+		return fmt.Errorf("failed to get Management: %w", err)
+	}
+
+	mgmtID := string(mgmt.UID)
+
+	templatesList := new(kcmv1.ClusterTemplateList)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.List(ctx, templatesList)
+	}); err != nil {
+		return fmt.Errorf("failed to list ClusterTemplates: %w", err)
+	}
+
+	tplName2Template := make(map[string]*kcmv1.ClusterTemplate, len(templatesList.Items))
```

Comment:
I think this part should be reworked as well after the comment https://github.com/k0rdent/kcm/pull/1783/files#r2247453352

What if there are two templates with the same name in different namespaces with different content?

Reply:
Thanks, [fixed](https://github.com/k0rdent/kcm/compare/2ee2e47dc247af8fb519346b98f5279a749752e6..51b53cc523304a1ceef243c257bc89055485c62c)